lib/src/query.dart

```diff
@@ -0,0 +1,766 @@
+    }finally {
+      listPtr.free();
+    }
+  }
+
+}
+
+class DoubleCondition extends Condition<double> {
+  DoubleCondition(ConditionOp op, ConditionType type, double value, double value2)
+      : super(op, type, value, value2);
+
+  int _op1(Pointer<Void> qbPtr, QueryCondition qc, obx_qb_cond_operator_1_dart_t<double> func) {
+    return func(qbPtr, qc._propertyId, _value);
+  }
+}
+
+/**
+ * The first element of the chain
+ * contains the enum representation of
+ * the to-be-constructed query builder.
+ * This design allows nested chains inside
+ * the chain.
+ */
+class QueryCondition {
+  bool _root = true;
+  int _entityId, _propertyId;
+  Condition _condition;
+  List<List<QueryCondition>> _dnf; // all
+  int _group = 1;
+
+  QueryCondition(this._entityId, this._propertyId, this._condition);
+
+  // && is not overridable
+  QueryCondition operator&(QueryCondition rh) => and(rh);
+
+  // || is not overridable
+  QueryCondition operator|(QueryCondition rh) => or(rh);
+
+  // TODO remove later
+  void debug() {
+    _dnf?.forEach((qc) => qc.map((c) => c._condition).forEach((c) => print("${c._value}")));
+    print("dnf size: ${_dnf.length}");
+  }
+
+  void _initDnfList(QueryCondition qc) {
+    _dnf ??= <List<QueryCondition>>[];
+    if (_dnf.length < _group) {
+      _dnf.add(<QueryCondition>[]);
+    }
+    _dnf[_group - 1].add(qc);
+  }
+
+  QueryCondition or(QueryCondition rh) {
+    rh._root = false;
+    if (_dnf == null) {
+      _initDnfList(this);
+    }
+    _group++;
+    _initDnfList(rh);
+    return this;
+  }
+
+  QueryCondition and(QueryCondition rh) {
+    rh._root = false;
+    if (_dnf == null) {
+      _initDnfList(this);
+    }
+    _initDnfList(rh);
+    return this;
+  }
+}
+
+class Query<T> {
+  Pointer<Void> _query;
+  Box<T> _box;
+
+  // package private ctor
+  Query._(Box<T> box, Pointer<Void> qb) {
+    _box = box;
+    _query = checkObxPtr(bindings.obx_query_create(qb), "create query", true);
+  }
+
+  int count() {
+    final ptr = Pointer<Uint64>.allocate(count: 1);
+    try {
+      checkObx(bindings.obx_query_count(_query, ptr));
+      return ptr.load();
+    }finally {
+      ptr.free();
+    }
+  }
+
+  // TODO does dart have a dtor/finalizer?j
+  void close() {
+    checkObx(bindings.obx_query_close(_query));
+  }
+
+  // TODO reimplement the same marshalling as in Box.getMany???
+  T findFirst() {
+    final list = findIds(offset:0, limit:1);
+    return (list == null ? null : _box.get(list.first)) as T;
+  }
+
+  List<int> findIds({int offset=0, int limit=(1<<32)}) {
+    final structPtr = checkObxPtr(bindings.obx_query_find_ids(_query, offset, limit), "find ids");
+    final idArray = IDArray.fromAddress(structPtr.address);
+    return idArray.ids.length == 0 ? null : idArray.ids;
+  }
+
+  List<T> find({int offset=0, int limit=(1<<32)}) {
+    final list = findIds(offset:offset, limit:limit);
+    return list == null ? null : _box.getMany(list);
```

Comment:
this way it's a little inefficient so if anything, it should be done in a read transaction & be marked with a big TODO to use `obx_query_find()`

Reply:
I totally agree.

Reply:
Done.